### ISSUE TITLE
Shared Theme Link overrides present theme by default theme.

### ISSUE THREAD

**Issue by ankiiitraj** [mozilla/FirefoxColor#866]
If you open link of shared theme which you are already using, the present theme is replaced by some other default theme.

Steps to reproduce:
- Open color.firefox.com and click on share then copy the link.
- Open the copied link 

![Peek 2019-11-12 21-32](https://user-images.githubusercontent.com/48787278/68690516-26a90380-0598-11ea-9fea-88e9f38b3071.gif)

**Comment by ankiiitraj** [comment]
I want to solve this issue, Can anyone please provide me some hints.

**Comment by caitmuenster** [comment]
Thanks @ankiiitraj, this one is interesting. @Rob--W, is this issue appropriate for a `good-first-bug` label?

**Comment by anuragakella** [comment]
Hey, I'd like to contribute to this. Can someone guide through?

**Comment by ankiiitraj** [comment]
Hi I'm not current working on this you can take on. Thanks

**Comment by anuragakella** [comment]
> 
> 
> Hi I'm not current working on this you can take on. Thanks

Hey, cool I'd love to work on it, but can someone guide me? @Rob--W, @caitmuenster ..

**Comment by Rob--W** [comment]
This was not marked as good first bug, but if you're willing to debug the issue I'm willing to mentor it. The first step is to set up a development environment (the README shows how to do that). After that you can dive into this bug.

The web page is designed to work with and without extension. To communicate between the web page and the extension, the `window.postMessage` API is used. When the extension is not installed, the web page will work and keep all data locally. When the extension *is* installed, then the extension will receive messages from the page (e.g. setTheme, fetchTheme, ping) and also reply (e.g. fetchedTheme, pong).

I was not able to consistently reproduce this bug, but when it occurred, the "setTheme" message was sent by the web page before "fetchTheme":

- "setTheme" (in response to "pong" from the extension, which is in response to "ping" from the web page) - https://github.com/mozilla/FirefoxColor/blob/a43d5c66c7dc6f644f4e42e6a53ed1df9e19f099/src/web/index.js#L157
- "fetchTheme" - https://github.com/mozilla/FirefoxColor/blob/a43d5c66c7dc6f644f4e42e6a53ed1df9e19f099/src/web/index.js#L300

Obviously, when the web page sends a `setTheme` message before the extension had a chance to respond to `fetchTheme`, then the web page does not know about the browser's actual theme, and the extension will receive the default theme (and overwrite the extension's actual theme with that default theme).

I think that in order to fix this bug, that a flag should be added to this place, so that the extension can ignore the message if a theme has already been installed: https://github.com/mozilla/FirefoxColor/blob/a43d5c66c7dc6f644f4e42e6a53ed1df9e19f099/src/web/index.js#L153-L157

**Comment by anuragakella** [comment]
Hey, thanks for the response!

Yes, I already built the code when I first saw this bug, and I was able to get it to run locally.
My first idea was to fetch the current theme and re-set it, every time the front end is loaded.
But your idea sounds better!

I managed to consistently reproduce the bug, [10/10 times], using these steps:
1. Set a theme from the theme presets
2. Copy the share link for the theme that was set in step 1
3. Set another theme from the presets, just to make sure the difference is visible.
4. Close the tab.
5. Wait for 10s and open a new tab, and paste the share link.

When I followed those steps, I was able to reproduce it every time, and the theme resets to a 'blue-pink' default.
I don't understand why, but it only happens if I wait for atleast 5-10sec after closing the current Firefox Color tab. If I paste the link immediately, the theme stays, i.e, no bug [sometimes].

Yes that was the first thing I observed too:
![image](https://user-images.githubusercontent.com/47664804/77610274-3e31c900-6f48-11ea-8b09-69ba10eae643.png)
Its doing a 'setTheme' before a 'fetchTheme' which doesn't make sense at all.

I was stuck here, until you explained about the postMessage API
I'll try and add a flag to the snippet you mentioned.

Thanks a lot for mentoring my first bug here!

**Comment by anuragakella** [comment]
Hey, @Rob--W I think I fixed the bug!

For the code here:
https://github.com/mozilla/FirefoxColor/blob/a43d5c66c7dc6f644f4e42e6a53ed1df9e19f099/src/web/index.js#L153-L157

First, I followed your suggestion and after a few changes I added an if condition:
Now the same lines look like:
```
if (!params.theme) {
  const state = store.getState();
  postMessage("addImages", {
    images: selectors.themeCustomImages(state)
  });
  postMessage("setTheme", { theme: selectors.theme(state) });
}
```
Since the bug occurs only when a shared link is opened, I decided to add a flag and check if the current page was a shared link or not using `params.theme` to check for ?theme, and it works!
if the page IS a shared link, then I added another `postMessage("fetchTheme");` here at line 281, the first line in the else block:
https://github.com/mozilla/FirefoxColor/blob/a43d5c66c7dc6f644f4e42e6a53ed1df9e19f099/src/web/index.js#L280-L290

I did that because, if the theme was not a shared link, the existing code already did a fetchTheme and worked properly. But if `params.theme` exists, there was no fetchTheme call.

Is this a good way to fix it?
Is there a better method?
Now, is there anything I should do before sending in a pull request?

Thanks!

**Comment by Rob--W** [comment]
FYI the presence of this bug was blocking work in PR #844, so this bug is going to be resolved by #844 . For those who are curious, the issue and the proposed fix are described in https://github.com/mozilla/FirefoxColor/pull/844#pullrequestreview-400425546

**Comment by Rob--W** [comment]
Fixed by #844.
